### PR TITLE
Update GettingStarted.md

### LINKED ISSUE #547: Add documentation for ManagedStream
Update getting started documentation to take advantage of the easiest way to begin using Oboe.

### PR DESCRIPTION
**Pull request by atneya**
Updating GettingStarted.md to contain documentation describing the usage of ManagedStream, as well as an extremely simple implementation of an Oboe stream. Fixes #547

### CODE REVIEW

File: docs/GettingStarted.md [hunk 1]
```diff
@@ -1,15 +1,29 @@
...
 **Make a note of the path which you cloned oboe into - you will need it shortly**
 
+If you use git as a VCS, consider adding Oboe as a submodule (underneath your
+app directory)
+
+```
+git submodule add https://github.com/google/oboe
+```
+
+This makes it easier to integrate updates to Oboe into your app.
```
Comment (atneya) [inline docs/GettingStarted.md]: Also makes developing for Oboe easier.

File: docs/GettingStarted.md [hunk 2]
```diff
@@ -119,23 +141,116 @@ Check the properties of the created stream. The **format** is one property which
...
+```
+oboe::ManagedStream managedStream;
+```
+Open the ManagedStream.
+```
+builder.openManagedStream(managedStream);
+```
+Start the ManagedStream in order to cause it to begin calling back.
+```
+managedStream->requestStart(); // We forgot to define a callback!
```
Comment (atneya) [inline docs/GettingStarted.md]: Separate reconfig/copy constructor from use of ManagedStream.

File: docs/GettingStarted.md [hunk 3]
```diff
@@ -119,23 +141,116 @@ Check the properties of the created stream. The **format** is one property which
...
+```
+In order to change the configuration of the stream, simply call this method
+again. The existing stream is closed, destroyed and a new stream is built and
+populates the managedStream.
+```
+// Create a builder with the existing stream properties
+builder = oboe::AudioStreamBuilder(*managedStream);
+// Re-open the stream with some additional config
+// The old ManagedStream is automatically closed and deleted
+builder.setCallback(mCallback)->openManagedStream(managedStream);
```
Comment (atneya) [inline docs/GettingStarted.md]: Avoid chaining

File: docs/GettingStarted.md [hunk 4]
```diff
@@ -119,23 +141,116 @@ Check the properties of the created stream. The **format** is one property which
...
+
+The following class is a complete implementation of a ManagedStream, which
+renders a sine wave. Creating the class (e.g. through the JNI bridge) creates
+and opens an Oboe stream which renders audio, and its destruction stops and
+closes the stream.
+```
+#include <oboe/Oboe.h>
+#include <math.h>
+
+class MyAudioEngine : public oboe::AudioStreamCallback {
```
Comment (atneya) [inline docs/GettingStarted.md]: Rename to SineRenderer

File: docs/GettingStarted.md [hunk 5]
```diff
@@ -119,23 +141,116 @@ Check the properties of the created stream. The **format** is one property which
...
+public:
+
+
+    MyAudioEngine() {
+        oboe::AudioStreamBuilder builder;
+        builder.setSharingMode(oboe::SharingMode::Exclusive)->setPerformanceMode(oboe::PerformanceMode::LowLatency)
+                ->setChannelCount(kChannelCount)->setSampleRate(kSampleRate)->setFormat(oboe::AudioFormat::Float)
```
Comment (atneya) [inline docs/GettingStarted.md]: Avoid chaining

File: docs/GettingStarted.md [hunk 6]
```diff
@@ -119,23 +141,116 @@ Check the properties of the created stream. The **format** is one property which
...
+                ->setCallback(this)->openManagedStream(outStream);
+        outStream->requestStart();
+    }
+
+    oboe::DataCallbackResult onAudioReady(oboe::AudioStream *oboeStream, void *audioData, int32_t numFrames) override {
+        float *floatData = (float *) audioData;
+        for (int i = 0; i < numFrames; ++i) {
+            for (int j = 0; j < kChannelCount; j++) {
+                floatData[i * kChannelCount + j] = kAmplitude * sinf(mPhase);
```
Comment (atneya) [inline docs/GettingStarted.md]: Pull out of inner loop

File: docs/GettingStarted.md [hunk 7]
```diff
@@ -119,23 +141,116 @@ Check the properties of the created stream. The **format** is one property which
...
+            if (mPhase >= kTWOPI) mPhase -= kTWOPI;
+        }
+        return oboe::DataCallbackResult::Continue;
+    }
+
+private:
+    oboe::ManagedStream outStream;
+    // Stream params
+    static int constexpr kChannelCount = 2;
+    static int constexpr kSampleRate = 48000;
```
Comment (atneya) [inline docs/GettingStarted.md]: Add comments about making member vars.

File: docs/GettingStarted.md [hunk 8]
```diff
@@ -119,23 +141,116 @@ Check the properties of the created stream. The **format** is one property which
...
+    // Wave params
+    static float constexpr kAmplitude = 0.5f;
+    static int constexpr kFrequency = 440;
```
Comment (atneya) [inline docs/GettingStarted.md]: Double

File: docs/GettingStarted.md [hunk 9]
```diff
@@ -119,23 +141,116 @@ Check the properties of the created stream. The **format** is one property which
...
+    static float constexpr kTWOPI = kPI * 2;
+    static double constexpr mPhaseIncrement = kFrequency * kTWOPI / (double) kSampleRate;
+    // Keeps track of where the wave is
+    float mPhase = 0.0;
+};
+```
+Note that this implementation computes  sine values at run-time for simplicity,
+rather than pre-computing them.
+Additionally, best practice is to implement a seperate callback class, rather
+than managing the stream and defining its callback in the same class.
```
Comment (atneya) [inline docs/GettingStarted.md]: Don't need to open stream immediately

File: docs/GettingStarted.md [hunk 10]
```diff
@@ -119,23 +141,116 @@ Check the properties of the created stream. The **format** is one property which
...
+
```
Comment (atneya) [inline docs/GettingStarted.md]: Query stream params before start

File: docs/GettingStarted.md [hunk 11]
```diff
@@ -119,23 +141,116 @@ Check the properties of the created stream. The **format** is one property which
...
+For more examples on how to use `ManagedStream` look in the `samples` folder.
+`samples/shared` contains an `AudioEngine` which can be easily inherited to begin
+creating interactive Oboe streams.
```
Comment (atneya) [inline docs/GettingStarted.md]: More info about this.

File: docs/GettingStarted.md [hunk 12]
```diff
@@ -1,15 +1,30 @@
 # Getting Started
 The easiest way to start using Oboe is to build it from source by adding a few steps to an existing Android Studio project.
 
+## Creating an Android app with native support
++ Create a new project by navigating `File > New > New Project`
```
Comment (dturner) [inline docs/GettingStarted.md]: Create a new project: `File > New > New Project`

File: docs/GettingStarted.md [hunk 13]
```diff
@@ -1,15 +1,30 @@
...
 
+If you use git as a VCS, consider adding Oboe as a submodule (underneath your
+app directory)
+
+```
+git submodule add https://github.com/google/oboe
+```
+
+This makes it easier to integrate updates to Oboe into your app, as well as
+contribute to Open Source.
```
Comment (dturner) [inline docs/GettingStarted.md]: This makes it easier to integrate updates to Oboe into your app, as well as
contribute to the Oboe project

File: docs/GettingStarted.md [hunk 14]
```diff
@@ -119,23 +142,126 @@ Check the properties of the created stream. The **format** is one property which
...
-It is important to close your stream when you're not using it to avoid hogging audio resources which other apps could use. To do this use: 
+It is important to close your stream when you're not using it to avoid hogging audio resources which other apps could use. To do this use:
 
     stream->close();
 
 `close()` is a blocking call which also stops the stream.
 
 It is usually advisable to close your stream when [`Activity.onPause()`](https://developer.android.com/guide/components/activities/activity-lifecycle#onpause) is called.
 
+## Using a ManagedStream
```
Comment (dturner) [inline docs/GettingStarted.md]: I think for the getting started guide we should recommend a single approach: either a `ManagedStream` or an `AudioStream`. My preference is to use `ManagedStream` as this is simpler.

Additionally, the benefits of using a `ManagedStream` should be explained up front i.e. mentioning that it handles sample rate, format and channel count conversions automatically etc.

File: docs/GettingStarted.md [hunk 15]
```diff
@@ -119,23 +142,126 @@ Check the properties of the created stream. The **format** is one property which
...
+};
+```
+Note that this implementation computes  sine values at run-time for simplicity,
+rather than pre-computing them.
+Additionally, best practice is to implement a seperate callback class, rather
+than managing the stream and defining its callback in the same class.
+This class also automatically starts the stream upon construction. Typically,
+the stream is queried for information prior to being started (e.g. burst size),
+and started upon user input.
+For more examples on how to use `ManagedStream` look in the `samples` folder.
```
Comment (dturner) [inline docs/GettingStarted.md]: Add link to samples

File: docs/GettingStarted.md (hunk 14 above)
Comment (atneya) [reply docs/GettingStarted.md]: In regards to the second paragraph, the ManagedStream does not handle conversions automatically, that would be added as part of the flowgraph changes, and would happen in either case. The primary benefits of a ManagedStream fall under memory and lifecycle management.

File: docs/GettingStarted.md [hunk 16]
```diff
@@ -1,15 +1,29 @@
...
-Start by cloning the [latest stable release](https://github.com/google/oboe/releases/) of the Oboe repository, for example: 
+### 1. Clone the github repository
+Start by cloning the [latest stable release](https://github.com/google/oboe/releases/) of the Oboe repository, for example:
 
-    git clone -b 1.1-stable https://github.com/google/oboe
+    git clone -b 1.2-stable https://github.com/google/oboe
 
 **Make a note of the path which you cloned oboe into - you will need it shortly**
 
+If you use git as a VCS, consider adding Oboe as a submodule (underneath your
```
Comment (philburk) [inline docs/GettingStarted.md]: c/as a VCS/as your version control system/

File: docs/GettingStarted.md [hunk 17]
```diff
@@ -1,15 +1,29 @@
...
+app directory)
```
Comment (philburk) [inline docs/GettingStarted.md]: Link the word "submodule" to this handy page:
https://gist.github.com/gitaarik/8735255

File: docs/GettingStarted.md [hunk 18]
```diff
@@ -119,23 +141,126 @@ Check the properties of the created stream. The **format** is one property which
...
 
 At this point you should start receiving callbacks.
 
 ### Closing the stream
-It is important to close your stream when you're not using it to avoid hogging audio resources which other apps could use. To do this use: 
+It is important to close your stream when you're not using it to avoid hogging audio resources which other apps could use. To do this use:
 
     stream->close();
 
 `close()` is a blocking call which also stops the stream.
```
Comment (philburk) [inline docs/GettingStarted.md]: Developers should call requestStop() before close().
Close() will generally stop the stream but it should not be relied upon.
There was a bug in MMAP code that prevented disconnected streams from getting stopped properly.

File: docs/GettingStarted.md [hunk 19]
```diff
@@ -119,23 +141,126 @@ Check the properties of the created stream. The **format** is one property which
...
 
 It is usually advisable to close your stream when [`Activity.onPause()`](https://developer.android.com/guide/components/activities/activity-lifecycle#onpause) is called.
```
Comment (philburk) [inline docs/GettingStarted.md]: This is particularly true when using SharingMode::Exclusive because you might prevent other apps from using the MMAP data path.

File: docs/GettingStarted.md [hunk 20]
```diff
@@ -119,23 +141,126 @@ Check the properties of the created stream. The **format** is one property which
...
+```
+// Modify the builder with some additional properties at runtime.
+builder.setDeviceId(MY_DEVICE_ID);
+// Re-open the stream with some additional config
+// The old ManagedStream is automatically closed and deleted
+builder.openManagedStream(managedStream);
+```
+The `ManagedStream` takes care of its own closure and destruction. If used in an
+automatic allocation context (such as a member of a class), the stream does not
+need to be closed or deleted. Make sure that the object which is responsible for
```
Comment (philburk) [inline docs/GettingStarted.md]: Does ManagedStream requestStop() before close()?

File: docs/GettingStarted.md [hunk 21]
```diff
@@ -119,23 +141,126 @@ Check the properties of the created stream. The **format** is one property which
...
+
+class OboeSinePlayer: public oboe::AudioStreamCallback {
+public:
+
+
+    OboeSinePlayer() {
+        oboe::AudioStreamBuilder builder;
+        // The builder set methods can be chained for convenience.
+        builder.setSharingMode(oboe::SharingMode::Exclusive)->setPerformanceMode(oboe::PerformanceMode::LowLatency);
+        builder.setChannelCount(kChannelCount)->setSampleRate(kSampleRate)->setFormat(oboe::AudioFormat::Float);
```
Comment (philburk) [inline docs/GettingStarted.md]: put each set() on its own line. I misread the code because of this.
You can still use continuation to chain the builder calls.

File: docs/GettingStarted.md [hunk 22]
```diff
@@ -119,23 +141,126 @@ Check the properties of the created stream. The **format** is one property which
...
+private:
+    oboe::ManagedStream outStream;
+    // Stream params
+    static int constexpr kChannelCount = 2;
+    static int constexpr kSampleRate = 48000;
+    // Wave params, these could be instance variables in order to modify at runtime
+    static float constexpr kAmplitude = 0.5f;
+    static float constexpr kFrequency = 440;
+    static float constexpr kPI = M_PI;
+    static float constexpr kTWOPI = kPI * 2;
```
Comment (philburk) [inline docs/GettingStarted.md]: I was about to suggest using M_2_PI but then I looked it up and found it equals 2/pi !! Egad!

Consider kTwoPi

File: docs/GettingStarted.md [hunk 23]
```diff
@@ -39,22 +58,25 @@ Here's a complete example `CMakeLists.txt` file:
...
+
     # Make the Oboe public headers available to our app
     include_directories (${OBOE_DIR}/include)
 
-Now go to `Build->Refresh Linked C++ Projects` to have Android Studio index the Oboe library. 
+    # Specify the libraries which our native library is dependent on, including Oboe
+    target_link_libraries (native-lib log oboe)
+
+
+Make sure that the `CMakeLists.txt` file lives next to your module specific build.gradle.
```
Comment (atneya) [inline docs/GettingStarted.md]: Change location

**Comment by dturner** [comment]
@atneya Please update to use ManagedStream

**Comment by philburk** [comment]
Also get Don's approval.

File: docs/GettingStarted.md [hunk 24]
```diff
@@ -73,6 +95,18 @@ Use the builder's set methods to set properties on the stream (you can read more
     builder.setFormat(oboe::AudioFormat::Float);
     builder.setChannelCount(oboe::ChannelCount::Mono);
 
+The builder's set methods can be easily chained:
```
Comment (dturner) [inline docs/GettingStarted.md]: I don't like chaining because IMO it hinders readability and looks messy. Could just be me.

File: docs/GettingStarted.md [hunk 25]
```diff
@@ -73,6 +95,18 @@ Use the builder's set methods to set properties on the stream (you can read more
...
+
+```
+oboe::AudioStreamBuilder builder;
+
+// Make sure to  always set the Performance and Sharing mode
+builder.setPerformanceMode(oboe::PerformanceMode::LowLatency)
+  ->setSharingMode(oboe::SharingMode::Exclusive); // The builder setter methods
+  can be chained for convenience
+builder.setCallback(myCallback)->setFormat(oboe::AudioFormat::Float);
```
Comment (dturner) [inline docs/GettingStarted.md]: Definitely don't chain on a single line though - that is positively horrible!

File: docs/GettingStarted.md [hunk 26]
```diff
@@ -18,48 +32,56 @@ Open your app's `CMakeLists.txt`. This can be found under `External Build Files`
...
 
 
-In the same file find the [`target_link_libraries`](https://cmake.org/cmake/help/latest/command/target_link_libraries.html) command. 
+In the same file find the [`target_link_libraries`](https://cmake.org/cmake/help/latest/command/target_link_libraries.html) command.
 Add `oboe` to the list of libraries which your app's library depends on. For example:
 
     target_link_libraries(native-lib oboe)
 
-Here's a complete example `CMakeLists.txt` file:
+Here's a complete example `CMakeLists.txt` file (without `OBOE_DIR` set):
```
Comment (dturner) [inline docs/GettingStarted.md]: Why did you add "without OBOE_DIR set"? What are you trying to tell the reader?

Comment (atneya) [reply docs/GettingStarted.md]: OBOE_DIR needs to be set for this build to compile. Right now it is simply a undefined variable.

File: docs/GettingStarted.md (hunk 24 above)
Comment (atneya) [reply docs/GettingStarted.md]: We should at least alert developers to the option of chaining if they wish. I think it then is up to developer preference.

File: docs/GettingStarted.md (hunk 26 above)
Comment (dturner) [reply docs/GettingStarted.md]: It's set a few lines below `set (OBOE_DIR ./oboe)`
